astropy/table/index.py

```diff
@@ -0,0 +1,308 @@
+        return self.data.range(lower, upper, bounds)
+
+    def replace(self, row, col, val):
+        self.remove_row(row, reorder=False)
+        key = [c[row] for c in self.columns]
+        key[self.col_position(col)] = val
+        self.data.add(tuple(key), row)
+
+    def replace_rows(self, col_slice):
+        row_map = dict((row, i) for i, row in enumerate(col_slice))
+        self.data.replace_rows(row_map)
+
+    def sorted_data(self):
+        return self.data.sort()
+
+    def __getitem__(self, item):
+        # item must be a slice; return sliced copy
+        return SlicedIndex(self, item)
+
+    def __str__(self):
+        return str(self.data)
+
+    def __repr__(self):
+        return str(self)
+
+    def __deepcopy__(self, memo):
+        # deep copy must be overridden to perform a shallow copy of columns
+        num_cols = self.data.num_cols if self.engine == SortedArray else None
+        index = Index(None, impl=self.data.__class__, col_dtypes=[x.dtype for
+                                                    x in self.columns])
+        index.data = deepcopy(self.data, memo)
+        index.columns = self.columns[:] # new list, same columns
+        memo[id(self)] = index
+        return index
+
+class SlicedIndex:
+    def __init__(self, index, index_slice):
+        if False: ##TODO index.engine == SortedArray:
+            self.index = Index(index.columns, index.data[index_slice])
+        else:
+            self.index = index
+
+        num_cols = len(index.columns[0])
+        self.start, self.stop, self.step = index_slice.indices(num_cols)
+
+    def sliced_coords(self, rows):
+        return [(row - self.start) / self.step for row in rows
+                if self.start <= row < self.stop and
+                (row - self.start) % self.step == 0]
+
+    def orig_coords(self, row):
+        return self.start + row * self.step
+
+    def find(self, key):
+        return self.sliced_coords(self.index.find(key))
+
+    def where(self, col_map):
+        return self.sliced_coords(self.index.where(col_map))
+
+    def range(self, lower, upper):
+        return self.sliced_coords(self.index.range(lower, upper))
+
+    def same_prefix(self, key):
+        return self.sliced_coords(self.index.same_prefix(key))
+
+    def sorted_data(self):
+        return self.sliced_coords(self.index.sorted_data())
+
+    def replace(self, row, col, val):
+        return self.index.replace(self.orig_coords(row), col, val)
+
+    ##TODO: adapt other Index methods here
+
+
+def get_index(table, table_copy):
+    '''
+    Inputs a table and some subset of its columns, and
+    returns an index corresponding to this subset or None
+    if no such index exists.
+    '''
+    cols = set(table_copy.columns)
+    indices = set()
+    for column in cols:
+        for index in table[column].indices:
+            if set([x.name for x in index.columns]) == cols:
+                return index
+    return None
+
+class static_indices:
```

Comment:
Given the name, I would have thought that this prevents the updating of indices as the values change (and then recalculates the index when the `with` block is exited).  Is that something that would be hard to implement?  The use case being "I need to add 1000 rows to a table, and don't want the index recalculated with each addition".  

Even if we don't implement what I just described, maybe we should change the name of this to avoid confusion?